### DISCUSSION TITLE
How to wrap a function around the signature pads

### DISCUSSION THREAD

**Discussion by TP-Harry-Phillips** [szimek/signature_pad#641] [Q&A] [answered]
What I am trying to achieve:
1. Have multiple empty divs on the page
2. Have a JS function that turns those divs into signature pads
3. Within the div are buttons that clear the pad and another that saves the content

I have been able to achieve 1 and 2 here:
[https://codesandbox.io/s/youthful-torvalds-eshp7d?file=/js/signPadApp.js](https://codesandbox.io/s/youthful-torvalds-eshp7d?file=/js/signPadApp.js)

I can't figure out how to do number 3.

First is the function **initSignaturePad**, this creates all the bits I want and adds them to the empty div, this works fine.

Next are the functions **clearPad** and **readAsSVG**, these accept an ID as a parameter. What I want them to do is select the appropriate pad and clear / save it. I don't know enough about JS to get it to work.

Is creating functions the correct way to achieve my goal?

I also can't figure out how to calculate the ratio and clear the pad on window resize either.

**Comment by UziTech** [comment]
instead of an `onclick` attribute you should be adding a listener to the `click` event

```diff
- v_clearButton.setAttribute("onclick", 'clearPad("' + vp_DivID + '");');
+ v_clearButton.addEventListener("click", () => { v_signaturePad.clear(); });
```

and similar with the `v_saveButton`

**Reply by TP-Harry-Phillips** [reply]
That works, in the sandbox, but it does not work on my test page.

I know there are differences in the sandbox and the test page such as the line:
`import SignaturePad from "signature_pad";`

I had to make those differences to get the sandbox to work, I did not think they would make a significant difference to the functionality of the code working or not working.

I changed that top line to:
`import SignaturePad from "./signature_pad.umd.js";`

In the sandbox that line works just fine, on my test page I am told that:
`import declarations may only appear at top level of a module`

I checked the error and all I found was pages that said if you want to import a function or class the source file needs to use:
`export class blahblahblah {`

**signature_pad.umd.js** does not have any exporting so I am pretty sure that the sandbox environment is doing some magic to make it work. I am stuck now, the sandbox only works because it adds something in the background and my own pages do not have that missing piece.

**Reply by TP-Harry-Phillips** [reply]
Ok , I just inspected the elements in the sandbox and the test page and found this:

Sandbox clear button:
<img width="403" alt="Screenshot_1597" src="https://user-images.githubusercontent.com/110009607/182048092-010b598a-45f7-419a-a495-4333d138406b.png">

Test page clear button:
<img width="398" alt="Screenshot_1598" src="https://user-images.githubusercontent.com/110009607/182048102-1e2bcdb6-8a88-4eeb-b68e-e9dd867e482a.png">

**Reply by TP-Harry-Phillips** [reply]
Nope not it, the test page now looks like this and left clear button still clears the right pad
<img width="395" alt="Screenshot_1599" src="https://user-images.githubusercontent.com/110009607/182048220-45dafde3-39e3-46f5-8be6-03a9e6d6553e.png">
:

**Reply by TP-Harry-Phillips** [reply]
When I add this line:
`console.log(v_signaturePad);`

This is what I get from clicking the left and right clear buttons:
<img width="471" alt="Screenshot_1600" src="https://user-images.githubusercontent.com/110009607/182051308-fd841a2b-54d2-4ebb-9d59-6374dc42c900.png">

Somehow the sandbox system keeps the two objects separate but my test page does not:
<img width="419" alt="Screenshot_1601" src="https://user-images.githubusercontent.com/110009607/182051363-87876ccf-0980-426a-b8c8-bfc4ebb6477d.png">

**Reply by UziTech** [reply]
You can check out our example at https://github.com/szimek/signature_pad/blob/master/docs/js/app.js. I suggest reading up on [closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures).

**Comment by TP-Harry-Phillips** [comment, marked as answer]
Ah ha! After looking at your example page again I figured out that in the function `initSignaturePad` I simply needed to use `const` at the front of the line:
`v_signaturePad = new SignaturePad(v_canvas, {backgroundColor: "rgb(0, 0, 0, 0)"});`

Now all the buttons reference the correct object.
<img width="554" alt="Screenshot_1602" src="https://user-images.githubusercontent.com/110009607/182054290-26098be5-3504-4856-857f-9eccb956d9df.png">
<img width="492" alt="Screenshot_1603" src="https://user-images.githubusercontent.com/110009607/182054537-0582bcbe-cb57-4d3d-a9cb-7fa7746f59fc.png">
